File: warp-drive/Messaging/WebSocketClient.cpp
#include "warpr_includes.h"
#include "WebSocketClient.h"

using namespace Axodox::Infrastructure;
using namespace Axodox::Json;
using namespace Axodox::Threading;
using namespace std;
using namespace std::string_literals;

namespace Warpr::Messaging
{
  WebSocketClient::WebSocketClient(Axodox::Infrastructure::dependency_container* container) :
    Connected(_events),
    MessageReceived(_events),
    Disconnected(_events),
    _settings(container->resolve<WarpConfiguration>())
  {
    //Connected
    _socket.onOpen([=] {
      _logger.log(log_severity::information, "Connected to {}.", _settings->GatewayUri);

      ConnectionRequest connectionRequest;
      *connectionRequest.SessionId = _settings->SessionId;
      SendMessage(connectionRequest);

      _events.raise(Connected, this);
      });

    //Received
    _socket.onMessage([=](const auto& message) {
      ReceiveMessage(message);
      });

    //Disconnected
    _socket.onClosed([=] {
      _logger.log(log_severity::information, "Disconnected from {}.", _settings->GatewayUri);
      _events.raise(Disconnected, this);
      if (_isDisposing) return;

      this_thread::sleep_for(1s);
      _socket.open(_settings->GatewayUri);
      });

    //Start
    _socket.open(_settings->GatewayUri);
  }

  WebSocketClient::~WebSocketClient()
  {
    _isDisposing = true;
  }

  bool WebSocketClient::IsConnected() const
  {
    return _socket.isOpen();
  }

  void WebSocketClient::SendMessage(const WarprSignalingMessage& warprMessage)
  {
    std::string jsonMessage = stringify_json(&warprMessage);
    _socket.send(jsonMessage);
  }

  void WebSocketClient::ReceiveMessage(const rtc::message_variant& rawMessage)
  {
    if (!holds_alternative<string>(rawMessage))
    {
      _logger.log(log_severity::warning, "Expected text message, but received binary.");
      return;
    }

    auto& jsonMessage = get<string>(rawMessage);
    auto warprMessage = try_parse_json<unique_ptr<WarprSignalingMessage>>(jsonMessage);
    if (!warprMessage)
    {
      _logger.log(log_severity::error, "Failed to parse message:\n{}", jsonMessage);
      return;
    }

    _events.raise(MessageReceived, this, warprMessage->get());
  }
}

File: warp-client/projects/warpr-lib/src/lib/networking/web-socket-client.ts
import { EventOwner, EventPublisher, EventArgs } from "../infrastructure/events";
import { IMessagingClient } from "./messaging-client";

export class WebSocketClient<TMessage> implements IMessagingClient<TMessage> {
  private readonly _events = new EventOwner();

  public readonly Connected = new EventPublisher<IMessagingClient<TMessage>, EventArgs>(this._events);
  public readonly MessageReceived = new EventPublisher<IMessagingClient<TMessage>, TMessage>(this._events);
  public readonly Disconnected = new EventPublisher<IMessagingClient<TMessage>, EventArgs>(this._events);

  private _socket?: WebSocket;

  constructor(private _uri: string) { }

  Connect(): void {
    console.log(`Connecting to ${this._uri}...`);

    let socket = new WebSocket(this._uri);
    socket.onopen = () => this.OnConnected();
    socket.onmessage = (e: MessageEvent<string>) => this.OnMessageReceived(e);
    socket.onclose = () => this.OnDisconnected();
    socket.onerror = (e: Event) => this.OnError(e);
    this._socket = socket;
  }

  SendMessage(message: TMessage): void {
    if (!this._socket) return;

    let content = JSON.stringify(message);
    console.debug(`Sending message: ${content}`);

    this._socket.send(content);
  }

  protected OnConnected(): void {
    console.log(`Connected to: ${this._uri}.`);
    this._events.Raise(this.Connected, this, EventArgs.Empty);
  }

  protected OnMessageReceived(eventArgs: MessageEvent<string>): void {
    console.debug('Message received: ' + eventArgs.data);

    let message = JSON.parse(eventArgs.data) as TMessage;
    this._events.Raise(this.MessageReceived, this, message);
  }

  protected OnDisconnected(): void {
    console.log(`Disconnected from: ${this._uri}.`);
    this._events.Raise(this.Disconnected, this, EventArgs.Empty);
  }

  protected OnError(error: Event): void {
    console.log('Connection failed: ' + error);
  }
}
